### DISCUSSION TITLE
Cartesi as Generalized Optimistic Rollup

### DISCUSSION THREAD

**Discussion by open-contracts** [cartesi/cips#10] [Ideas]
I want to propose and discuss integrating Cartesi more deeply with an existing Optimistic EVM Rollup design. In particular, [Optimism ](https://www.optimism.io/) might be a great candidate because the user base is growing quickly, it does not yet have an active fraud proving mechanism, and is [explicitly not commited to any particular fraud proof solution](https://medium.com/ethereum-optimism/our-pragmatic-path-to-decentralization-cb5805ca43c1) yet. Otherwise one could always fork them, of course.

Here's the rough idea: 
  - compile whatever version of geth the Optimism team is using to run on a Cartesi VM
  - keep all of Optimisms existing infrastructure (RPC endpoints, L1<->L2 messaging, etc) in place
  - let an optimism sequencer (of which multiple could exist) commit to an update of the Cartesi VM (subsuming an l2 block update inside geth), which can be computed deterministically from (compressed) calldata published on L1
  - use Cartesi fraud proofs to punish invalid state updates, slashing deposits of the sequencer. The  [complexities surrounding "global consensus" as pursued by Arbitrum](https://developer.offchainlabs.com/docs/inside_arbitrum) could be avoided by relying on a local consensus (Cartesi's current approach) among "validators" (=addresses who are whitelisted to claim and prove fraud). Anyone could opt into becoming a validator for _future_ EVM updates by depositing a stake. Previous state updates could only be called out by address who were previously registered as fraud provers, to alleviate some of the (DDoS) complexities of a global consensus design.
  - **the best part**: deploy a special precompiled contract to the geth instance running in the Cartesi VM, which can be called from other contracts to launch and await the results from arbitrary bash scripts executed outside geth but inside the Cartesi VM, supporting custom input drives that are published to L1 (later, native Noether support could be added as cheaper alternative for larger drives).

Main benefits: 
- instant call+confirmation of generic Cartesi computations at no additional trust assumptions (from the POV of an L2 solidity contract)
- contract devs who want to offload contract computations into cartesi do not have to design their own local consensus anymore (if they don't need privacy)
- by reusing the same Cartesi VM for all contracts, auditors only need to focus on the content of custom input drives and bash scripts, increasing security

Main problems:
- not much, nothing has to be given up as the existing Cartesi product roadmap could exist+develop in parallel
- need to think through if a local consensus with permissionless opt-in is actually easier to design than a global one

**Comment by augustoteixeira** [comment]
Having an EVM solution compiled to RISC-V and running on our Rollups is definitely something we have been thinking about.

Here are a few comments on this idea:
- geth is not a very good option, since it does much more than EVM, so we should probably settle for something lightweight, such as the [revm](https://github.com/bluealloy/revm).
- We will suffer a performance penalty dues to "double emulation", the EVM is already a virtual machine and it will be executed inside a RISC-V architecture, which is itself emulated. We have to benchmark to see how bad the situation is.
- We are actively working on the microarchitecture (it has been mentioned in public [here](https://medium.com/cartesi/cartesis-new-and-accessible-roadmap-fc5b8b8dda4c)). Once this is finished, we have a hope to compile the EVM directly to the microarchitecture and avoid the double emulation problem above. But that requires more time.
- The idea of having "Linux pre-compiled contracts" is very interesting. Here at Cartesi we call it the Minotaur (because it is sort of a hybrid architecture).
- It is not very easy to adapt Optimism's aggregator/sequencer to our setup, because the way which we process inputs is quite different from theirs. But we are working actively on an aggregator now and it should work just fine for this solution.
- Another advantage that you didn't mention is that we can use standard compression libraries (like Brotli) to decompress the aggregated inputs arriving to the EVM. This would reduce the data availability requirements of our Rollups, making it cheaper to use.
- Pursuing this idea as a finalized product would put us more into the direct route of competition with Optimism and Arbitrum. Which can be seen as a good thing, but requires some business analysis that we are performing right now. A partnership could also be very valuable in this case (ping @erickmoura).

Thanks for the message, we are going to post updates on this line of research here.
